3.8/graphs-custom-pregel.md

```diff
@@ -0,0 +1,1511 @@
+  with pregel-id `to-pregel-vertex`.
+- `send-to-all-neighbors` send the value `value` to the accumulator `name` in
+  all neighbors reachable by an edge. Note that if there are multiple edges
+  from us to the neighbor, the value is sent multiple times.
+
+#### This Vertex
+
+```js
+["this-doc"]
+["this-vertex-id"]
+["this-unique-id"]
+["this-pregel-id"]
+["this-outdegree"]
+["this-outbound-edges-count"]
+["this-outbound-edges"]
+```
+
+- `this-doc` returns the document slice stored in vertex data.
+- `this-outdegree` returns the number of outgoing edges.
+- `this-outbound-edges-count` alias for `this-outdegree`.
+- `this-outbound-edges` returns a list of outbound edges of the form
+  ```json
+  {
+    "document": <edge-document>,
+    "to-pregel-id": <to-vertex-pregel-id>
+  }
+  ```
+- `this-vertex-id` returns the vertex document identifier.
+- `this-unique-id` returns a unique but opaque numeric value associated with
+   this vertex.
+- `this-pregel-id` returns a identifier used by Pregel to send messages.
+
+#### Miscellaneous
+
+- `["vertex-count"]` returns the number of vertices in the graph under
+  consideration.
+- `["global-superstep"]` the current superstep the algorithm is in.
+- `["phase-superstep"]` the current superstep the current phase is in.
+- `["current-phase"]` the current phase name.
+
+### Foreign calls in _Conductor_ context
+
+The following functions are only available when running in the Conductor
+context to coordinate phases and phase changes and to access and modify
+global accumulators.
+
+#### Phase Management
+
+```
+["goto-phase", phase]
+["finish"]
+```
+
+`goto-phase` sets the current phase to `phase`. `finish` finishes the Pregel
+computation.
+
+#### Global Accumulators
+
+```
+["global-accum-ref", name]
+["global-accum-set!", name, value]
+["global-accum-clear!", name]
+```
+
+`global-accum-ref`, `global-accum-set!`, `global-accum-clear!` like for
+accumulators but for global accumulators.
+
+### Foreign calls in _Custom Accumulator_ context
+
+The following functions are only available when running inside a custom accumulators.
+
+- `["parameters"]`
+   returns the object passed as parameter to the accumulator definition
+- `["current-value"]`
+   returns the current value of the accumulator
+- `["get-current-value"]`
+   returns the current value but calls the `getProgram` to do so.
+- `["input-value"]`
+   returns the _input value_. This is the value received as update in
+   `updateProgram`. Or the value the accumulator is set to in `setProgram`.
+- `["input-sender"]`
+   returns the _vertex-id_ of the sending vertex. This is only available in
+   `updateProgram`.
+- `["input-state"]`
+   return the input state for a merge operation. This is only available in
+   `aggregateStateProgram`.
+- `["this-set!", value]`
+   set the new value of the accumulator to `value`.
+- `["this-set-value!", value]`
+   set the new value of the accumulator but calls the `setProgram` to do so.
```

Comment:
Instead of those function I would like the `getProgram` and stuff to be functions that are called with parameters. And their return value is used then. This feels more natural.

Reply:
Moved the issue to internal tracking board. Closing this here.